In [1]:
# Módulos importantes
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

# Cargando y visualizando datos

In [2]:
# Cargamos el archivo 'datos/heart_disease.csv' en un dataframe
df=pd.read_csv('datos/wisconsin_data.csv')

In [3]:
# Mostrar los primeras filas
df.head()

patientId  clumpThickness  cellSize  CellShape  marginalAdhesion  \
0          1               5         1          1                 1   
1          2               5         4          4                 5   
2          3               3         1          1                 1   
3          4               6         8          8                 1   
4          5               4         1          1                 3   

   epithelialSize  bareNuclei  blandChromatin  normalNucleoli  mitoses  label  
0               2         1.0               3               1        1      0  
1               7        10.0               3               2        1      0  
2               2         2.0               3               1        1      0  
3               3         4.0               3               7        1      0  
4               2         1.0               3               1        1      0

In [4]:
# Encontrar los principales valores estadísticos 
df.describe()

patientId  clumpThickness    cellSize   CellShape  marginalAdhesion  \
count  699.000000      699.000000  699.000000  699.000000        699.000000   
mean   350.000000        4.417740    3.134478    3.207439          2.806867   
std    201.928205        2.815741    3.051459    2.971913          2.855379   
min      1.000000        1.000000    1.000000    1.000000          1.000000   
25%    175.500000        2.000000    1.000000    1.000000          1.000000   
50%    350.000000        4.000000    1.000000    1.000000          1.000000   
75%    524.500000        6.000000    5.000000    5.000000          4.000000   
max    699.000000       10.000000   10.000000   10.000000         10.000000   

       epithelialSize  bareNuclei  blandChromatin  normalNucleoli     mitoses  \
count      699.000000  667.000000      699.000000      699.000000  699.000000   
mean         3.216023    3.536732        3.437768        2.866953    1.589413   
std          2.214300    3.683795        2.438364        3.053634    1.715078   
min          1.000000    1.000000        1.000000        1.000000    1.000000   
25%          2.000000    1.000000        2.000000        1.000000    1.000000   
50%          2.000000    1.000000        3.000000        1.000000    1.000000   
75%          4.000000    7.000000        5.000000        4.000000    1.000000   
max         10.000000   10.000000       10.000000       10.000000   10.000000   

            label  
count  699.000000  
mean     0.344778  
std      0.475636  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      1.000000

In [5]:
# Mostrar las columnas
df.columns

Index(['patientId', 'clumpThickness', 'cellSize', 'CellShape',
       'marginalAdhesion', 'epithelialSize', 'bareNuclei', 'blandChromatin',
       'normalNucleoli', 'mitoses', 'label'],
      dtype='object')

In [6]:
# Mostrar el número de filas y columnas
df.shape

(699, 11)

In [7]:
#Quitar la columna patiendId
df=df.drop(columns='patientId')
df.head()

clumpThickness  cellSize  CellShape  marginalAdhesion  epithelialSize  \
0               5         1          1                 1               2   
1               5         4          4                 5               7   
2               3         1          1                 1               2   
3               6         8          8                 1               3   
4               4         1          1                 3               2   

   bareNuclei  blandChromatin  normalNucleoli  mitoses  label  
0         1.0               3               1        1      0  
1        10.0               3               2        1      0  
2         2.0               3               1        1      0  
3         4.0               3               7        1      0  
4         1.0               3               1        1      0

In [8]:
#De manera general checar si hay elementos 'NaN'
df.isnull().any()

clumpThickness      False
cellSize            False
CellShape           False
marginalAdhesion    False
epithelialSize      False
bareNuclei           True
blandChromatin      False
normalNucleoli      False
mitoses             False
label               False
dtype: bool

In [9]:
#Revisando cuantos elementos NaN existen
np.isnan(df).sum()

clumpThickness       0
cellSize             0
CellShape            0
marginalAdhesion     0
epithelialSize       0
bareNuclei          32
blandChromatin       0
normalNucleoli       0
mitoses              0
label                0
dtype: int64

In [10]:
#Quitar filas con elementos faltantes ('NaN')
df=df.dropna()
df.describe()

clumpThickness    cellSize   CellShape  marginalAdhesion  \
count      667.000000  667.000000  667.000000        667.000000   
mean         4.385307    3.100450    3.167916          2.811094   
std          2.795851    3.051182    2.982731          2.864350   
min          1.000000    1.000000    1.000000          1.000000   
25%          2.000000    1.000000    1.000000          1.000000   
50%          4.000000    1.000000    1.000000          1.000000   
75%          6.000000    4.500000    5.000000          4.000000   
max         10.000000   10.000000   10.000000         10.000000   

       epithelialSize  bareNuclei  blandChromatin  normalNucleoli     mitoses  \
count      667.000000  667.000000      667.000000      667.000000  667.000000   
mean         3.185907    3.536732        3.413793        2.796102    1.583208   
std          2.195733    3.683795        2.439599        3.000568    1.701497   
min          1.000000    1.000000        1.000000        1.000000    1.000000   
25%          2.000000    1.000000        2.000000        1.000000    1.000000   
50%          2.000000    1.000000        3.000000        1.000000    1.000000   
75%          4.000000    7.000000        4.000000        3.000000    1.000000   
max         10.000000   10.000000       10.000000       10.000000   10.000000   

            label  
count  667.000000  
mean     0.341829  
std      0.474679  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      1.000000

In [11]:
# Mostrar el número de filas y columnas sin NaN
df.shape

(667, 10)

In [12]:
# Mostrar el tipo de variables
df.dtypes

clumpThickness        int64
cellSize              int64
CellShape             int64
marginalAdhesion      int64
epithelialSize        int64
bareNuclei          float64
blandChromatin        int64
normalNucleoli        int64
mitoses               int64
label                 int64
dtype: object

# Preparando datos para Machine Learning

In [13]:
# Crear X (Quitar la columna o campo de "label")
X=df.drop(columns='label')
X.head()

clumpThickness  cellSize  CellShape  marginalAdhesion  epithelialSize  \
0               5         1          1                 1               2   
1               5         4          4                 5               7   
2               3         1          1                 1               2   
3               6         8          8                 1               3   
4               4         1          1                 3               2   

   bareNuclei  blandChromatin  normalNucleoli  mitoses  
0         1.0               3               1        1  
1        10.0               3               2        1  
2         2.0               3               1        1  
3         4.0               3               7        1  
4         1.0               3               1        1

In [14]:
# Crear y (Valor de de la columna "label")
y=df['label']
y

0      0
1      0
2      0
3      0
4      0
      ..
694    0
695    0
696    1
697    0
698    0
Name: label, Length: 667, dtype: int64

In [15]:
# Dividir los datos para entrenamiento y prueba, se elige muestra test del 80%
X_train, X_test, y_train, y_test=train_test_split (X,y, train_size=0.8)

In [16]:
# Revise  el tamaño de cada muestra
print( '80% ' ,df.shape[0]*0.8)
print( 'X_train ' ,X_train.shape[0])
print( 'y_train ' ,y_train.shape[0])
print( 'X_test ' ,X_test.shape[0])
print( 'y_test' ,y_test.shape[0])

80%  533.6
X_train  533
y_train  533
X_test  134
y_test 134


# Selección de modelo y realización de entrenamiento

## a) DecisionTree
### Con parámetros de afinación o tunning extras de: **min_samples_split** y **max_features**

In [17]:
## Módulos del Modelo: Arboles de decision 
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score

## DecisionTree ajustado en clase (max_depth y min_samples_leaf)

In [18]:
# Ajuste de los parametros max_depth y min_samples_leaf
scores_DT_c=[]
for i in range (1,8) :
    for j in range (1,8) :
        #print('max_depth = ', i, 'min_samples_leaf=', j)
        modelo_DT = DecisionTreeClassifier( random_state=1,  
                                  max_depth = i,
                                  min_samples_leaf= j)
        modelo_DT.fit(  X_train, y_train  )
        #print('Score=' , modelo_DT.score( X_test, y_test ))
        #print('_')
        scores_DT_c.append((modelo_DT.score(X_test,y_test),i,j))

In [19]:
scores_DT_c.sort(reverse=True)
scores_DT_c[:1]
#(Score,max_depth,min_samples_leaf)

[(0.9701492537313433, 7, 5)]

## DecisionTree nuevo ajuste (min_samples_split y max_features)

In [20]:
## Se conservan los mejores max_depth y min_samples_leaf hayados previamente
### Se itera sobre min_samples_split y max_features
scores_DT=[]
for i in range (2,16) :
    for j in range (1,15) :
        #print('min_samples_split = ', i, 'max_features=', j)
        modelo_DT = DecisionTreeClassifier( random_state=1,  
                                  max_depth = scores_DT_c[0][1],
                                  min_samples_leaf= scores_DT_c[0][2],
                                  min_samples_split=i,
                                  max_features=j)
        modelo_DT.fit(  X_train, y_train  )
        #print('Score=' , modelo_DT.score( X_test, y_test ))
        #print('__________________________________________')
        scores_DT.append((modelo_DT.score(X_test,y_test),i,j))

In [21]:
scores_DT.sort(reverse=True)
scores_DT[:1]
#(Score,min_samples_split,max_features)

[(0.9776119402985075, 13, 8)]

# b) RandomForestClassifier
### Con parámetros de afinación o tunning extras de: **max_depth**, **min_samples_leaf** y **min_samples_split**

In [22]:
# Módulo RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

## RandomForestClassifier ajustado en clase (max_depth y n_estimators)

In [23]:
scores_RF_c=[]
for i in range (1,8) :
    for j in range (1,8) :
        #print('max_depth = ', i, 'n.estimators=', j)
        modelo_RF = RandomForestClassifier( random_state=1,  
                                  max_depth = i,
                                  n_estimators= j)
        modelo_RF.fit(  X_train, y_train  )
        #print('Score=' , modelo_RF.score( X_test, y_test ))
        #print('_')
        scores_RF_c.append((modelo_RF.score(X_test,y_test),i,j))

In [24]:
scores_RF_c.sort(reverse=True)
scores_RF_c[:1]
#(Score,max_depth,n.estimators)

[(0.9776119402985075, 7, 7)]

## RandomForestClassifier nuevo ajuste (max_depth,min_samples_leaf y min_samples_split)

In [25]:
## Se conserva el mejor n_estimators hayado previamente
## Se itera sobre max_depth, min_samples_leaf y min_samples_split
scores_RF=[]
for j in range (1,15):
    for p in range (2,16):
        #print('max_depth = ', i, 'min_samples_leaf =', j,'min_samples_split =',p)
        modelo_RF = RandomForestClassifier( random_state=1,  
                              max_depth = scores_RF_c[0][1],
                              n_estimators= scores_RF_c[0][2],
                              min_samples_leaf=j,
                              min_samples_split=p)
        modelo_RF.fit(X_train, y_train)
        #print('Score=' , modelo_RF.score( X_test, y_test ))
        #print('__________________________________________')
        scores_RF.append((modelo_RF.score(X_test,y_test),j,p))

In [26]:
scores_RF.sort(reverse=True)
scores_RF[:1]
#(Score,min_samples_leaf,min_samples_split)

[(0.9850746268656716, 9, 15)]

# Resultados

| Modelo o clasificador   |  Eficiencia | Parametros de ajuste  |
|---                      |---          |---                    |
|  Decision Tree_clase    |97.0149%     |max_depth=7, min_samples_leaf=5, min_samples_split=NA ,max_features=NA|
|  Decision Tree_nuevo    |97.7611%     |max_depth=7, min_samples_leaf=5, min_samples_split=13, max_features=8|
|  Random Forest_clase    |97.7611%     |max_depth=7, n_estimators=7, min_samples_leaf=NA, min_samples_split=NA|
|  Random Forest_nuevo    |98.5074%             |max_depth=7, n_estimators=7, min_samples_leaf=9, min_samples_split=15|

# Conclusiones

Los nuevos ajustes mejoraron la eficiencia en ambos modelos, tal y como se muestra en la tabla anterior. Sin embargo, la mejora no muy significativa, en el caso de DecesionTree solo hubo mejora en algunas décimas de la eficiencia, y para RandomForest la mejora fue de casi de un 1% más.
Además, como las muestras de testeo y entrenamiento se seleccionan de manera aleatoria, cada vez que se compila toda la notebook, las eficiencias resultan diferentes. 
El hecho de agregar más parámetros de ajuste a los modelos sí mejora la eficiencia siempre y cuando dicho parámetro este optimizado y para hacerlo, hay que iterar hasta encontrar el valor óptimo. Lamentablemente, esto conlleva mucho tiempo de procesamiento y las mejoras convergen en un valor.